rocketmq-flink/src/main/java/org/apache/rocketmq/flink/RocketMQSink.java

```diff
@@ -0,0 +1,187 @@
+import org.apache.flink.streaming.api.functions.sink.RichSinkFunction;
+import org.apache.flink.streaming.api.operators.StreamingRuntimeContext;
+import org.apache.rocketmq.client.exception.MQClientException;
+import org.apache.rocketmq.client.producer.DefaultMQProducer;
+import org.apache.rocketmq.client.producer.SendCallback;
+import org.apache.rocketmq.client.producer.SendResult;
+import org.apache.rocketmq.common.message.Message;
+import org.apache.rocketmq.flink.common.selector.TopicSelector;
+import org.apache.rocketmq.flink.common.serialization.KeyValueSerializationSchema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The RocketMQSink provides at-least-once reliability guarantees when
+ * checkpoints are enabled and batchFlushOnCheckpoint(true) is set.
+ * Otherwise, the sink reliability guarantees depends on rocketmq producer's retry policy.
+ */
+public class RocketMQSink<IN> extends RichSinkFunction<IN> implements CheckpointedFunction {
+
+    private static final long serialVersionUID = 1L;
+
+    private static final Logger LOG = LoggerFactory.getLogger(RocketMQSink.class);
+
+    private transient DefaultMQProducer producer;
+    private boolean async; // false by default
+
+    private Properties props;
+    private TopicSelector<IN> topicSelector;
+    private KeyValueSerializationSchema<IN> serializationSchema;
+
+    private boolean batchFlushOnCheckpoint; // false by default
+    private List<Message> batchList;
+
+    public RocketMQSink(KeyValueSerializationSchema<IN> schema, TopicSelector<IN> topicSelector, Properties props) {
+        this.serializationSchema = schema;
+        this.topicSelector = topicSelector;
+        this.props = props;
+    }
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        Validate.notEmpty(props, "Producer properties can not be empty");
+        Validate.notNull(topicSelector, "TopicSelector can not be null");
+        Validate.notNull(serializationSchema, "KeyValueSerializationSchema can not be null");
+
+        producer = new DefaultMQProducer();
+        producer.setInstanceName(String.valueOf(getRuntimeContext().getIndexOfThisSubtask()));
+        RocketMQConfig.buildProducerConfigs(props, producer);
+
+        batchList = new LinkedList<>();
+
+        if (batchFlushOnCheckpoint && !((StreamingRuntimeContext) getRuntimeContext()).isCheckpointingEnabled()) {
+            LOG.warn("Flushing on checkpoint is enabled, but checkpointing is not enabled. Disabling flushing.");
+            batchFlushOnCheckpoint = false;
+        }
+
+        try {
+            producer.start();
+        } catch (MQClientException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void invoke(IN input, Context context) throws Exception {
+        Message msg = prepareMessage(input);
+
+        if (batchFlushOnCheckpoint) {
+            batchList.add(msg);
```

Comment:
Add action isn't protected by the lock `batchList`， is this a concurrent issue?

Reply:
Not an issue. Each RocketMQSink instance belong to a thread. We just need to sync `batchList ` when `snapshotState` invoked.

Reply:
Got it.